### PR TITLE
[ENH] Partially replace `pd.Int64Index` with `pd.Index`

### PR DESCRIPTION
**Pull request by khrapovs**
#### Reference Issues/PRs

Partially addresses #2332.

#### What does this implement/fix? Explain your changes.

I have replaced `pd.Int64Index` with `pd.Index` only in obvious and harmless places. More complicated like `VALID_INDEX_TYPES` in `sktime.utils.validation.series` are left for another ambitious PR.

#### Does your contribution introduce a new dependency? If yes, which one?

No

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [X] I've added unit tests and made sure they pass locally.
- [X] The PR title starts with either [ENH], [DOC] or [BUG] indicating wether the PR topic is related to enhancement, documentation or bug

### CODE REVIEW

**Comment by khrapovs** [comment]
> not _entirely_ harmless, is it? smile image

OMG, what did I got myself into! now I see why others abandoned it :D

**Review by fkiraly — CHANGES_REQUESTED** [review]
Take that, `Int64Index`!

In-principle, this look ok, *but* - we need to deprecate!

Why: suppose some poor user's highly valuable deployed pipelines are relying entirely on `Int64Index` for some reason. This PR would completely break it, wouldn't it?

Breakage occurs wherever we do input checks and the index is no longer allowed.
In those cases, we need to raise a warning (deprecation since 0.12, removal in 0.13), while being defensive in the sense of making sktime robust against the `pandas` deprecation...

In the many cases where we *create* an index, there is no deprecation necessary.

**Comment by khrapovs** [comment]
@fkiraly Yes, you are right. _But_, right now there is no way to say whether a user created the index using `pd.Index` or `pd.Int64Index`:
```
type(pd.Index([1, -1]))
<class 'pandas.core.indexes.numeric.Int64Index'>
type(pd.Int64Index([1, -1]))
<input>:1: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
<class 'pandas.core.indexes.numeric.Int64Index'>
```

Additionally, from https://pandas.pydata.org/docs/reference/api/pandas.Int64Index.html:
> Deprecated since version 1.4.0: In pandas v2.0 Int64Index will be removed and NumericIndex used instead. Int64Index will remain fully functional for the duration of pandas 1.x.

But `NumericIndex` does not even exist yet!

**Comment by fkiraly** [comment]
@khrapows, I see. That makes things difficult.
Is there any way to not make *us* break our interface contracts?
How about replacing checks for index *types* by uses of the `pandas` method `Index.is_integer`?

**Comment by khrapovs** [comment]
very good suggestion! I completely overlooked that such a method exists. replaced all usages of `is_numeric` with `is_integer`. thanks!

**Review by fkiraly — CHANGES_REQUESTED** [review]
Nice!
I think this could almost go in, but I'm concerned about the maintainability of the type check, where we now repeatedly do sth like `isinstance(obj, VALID_TYPES) or (isinstance(obj, Index) and obj.is_integer())`.

"Why" is hard to understand to someone who hasn't followed our discussion, and it's also repeated in multiple places across the codebase now.

I would hence recommend to bundle this in one function, perhaps
`is_valid_fh_index(obj, type="any")`, where `type` can be `"relative"` or `"absolute"` - what do you think?

**Comment by khrapovs** [comment]
@fkiraly 

Very good! Added two new functions:
- `sktime.utils.validation.series.is_in_valid_index_types`
- `sktime.utils.validation.series.is_integer_index`

The check for `RELATIVE_TYPES` or `ABSOLUTE_TYPES` happens only a couple of times in the single file `_fh.py`. A combined check for either valid index or integer index only happens just two times. I do not think it makes sense to over-engineer for such rare use cases. But if you insist of a function with `type` argument, I would be happy to write a couple more lines of code.

**Comment by fkiraly** [comment]
> But if you insist of a function with type argument, I would be happy to write a couple more lines of code.

No, I do not insist. I just wanted to bundle the logic in one place, which you did.

Once the tests pass, I'm happy to approve.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Looking at it again, this is not entirely what I meant - I meant we should bundle the logic "check in relative/absolute types plus check is integer index" in one function.

That's because previously it was one line, since the "check is integer index" was, previously, "check for certain type".

So, possibly, one function `is_relative_fh_type` which does `isin RELATIVE_TYPES or is_integer_index` and `is_absolute_fh_type` which does `isin ABSOULTE_TYPES or is_integer_index`?

**Comment by khrapovs** [comment]
pushed two new functions: `functions is_relative_fh_type` and `and is_absolute_fh_type`

**Comment by fkiraly** [comment]
😱 
image
what is this now - seems unrelated?

Maybe let's record this as a sporadic failure?

**Comment by khrapovs** [comment]
> scream image what is this now - seems unrelated?
> 
> Maybe let's record this as a sporadic failure?

Reported here: https://github.com/alan-turing-institute/sktime/issues/2368

File: sktime/forecasting/base/_fh.py
Comment (fkiraly) [inline sktime/forecasting/base/_fh.py:110]: this should be `is_relative_fh_type or is_absolute_fh_type` imo, no?

File: sktime/forecasting/base/tests/test_fh.py
Comment (fkiraly) [inline sktime/forecasting/base/tests/test_fh.py:156]: `is_relative_index_type or is_absolute_index_type`

File: sktime/utils/validation/series.py
Comment (fkiraly) [inline sktime/utils/validation/series.py:32]: I would actually change this to `isinstance(x, VALID_INDEX_TYPES) or is_integer_index(x)` and then use it everywhere

**Review by fkiraly — CHANGES_REQUESTED** [review]
Ok, one last change request:
* change `is_in_valid_index_types` as suggested to include the `is_integer_index` check
* move it to `_fh.py`, next to the two other functions
* then use it in the two other places where you check that manually

File: sktime/forecasting/base/_fh.py
Comment (fkiraly) [reply sktime/forecasting/base/_fh.py:110]: better probably `is_in_valid_index_types` once it´s changed (see below)

File: sktime/forecasting/base/tests/test_fh.py
Comment (fkiraly) [reply sktime/forecasting/base/tests/test_fh.py:156]: better probably `is_in_valid_index_types` once it´s changed (see below)

**Comment by fkiraly** [comment]
(I'm not insistent about my idea `is_valid_fh_index(obj, type="any")`, where `type` can be `"relative"` or `"absolute"` - also happy with three loose functions - but just wanted to bring it up again in case it seems more natural now)

File: sktime/forecasting/base/_fh.py
Comment (khrapovs) [reply sktime/forecasting/base/_fh.py]: done

File: sktime/forecasting/base/tests/test_fh.py
Comment (khrapovs) [reply sktime/forecasting/base/tests/test_fh.py]: done

File: sktime/utils/validation/series.py
Comment (khrapovs) [reply sktime/utils/validation/series.py]: done

**Comment by khrapovs** [comment]
> * change `is_in_valid_index_types` as suggested to include the `is_integer_index` check

done

> * move it to `_fh.py`, next to the two other functions

`is_in_valid_index_types` is very generic and does not have much connection to `ForecastingHorizon` other than it is used once to check fh values. I would prefer to keep it next to where `VALID_INDEX_TYPES` are defined. This function is used much more frequently by other modules.

> * then use it in the two other places where you check that manually

done

> (I'm not insistent about my idea `is_valid_fh_index(obj, type="any")`, where `type` can be `"relative"` or `"absolute"` - also happy with three loose functions - but just wanted to bring it up again in case it seems more natural now)

Yes, I remember about this proposal and even now would prefer to keep the functions as they are now.

**Review by fkiraly — APPROVED** [review]
I'd be happy to approve now, although I leave the merge decision to you after considering my non-blocking comment below.

>`is_in_valid_index_types` is very generic and does not have much connection to `ForecastingHorizon` other than it is used once to check fh values. I would prefer to keep it next to where `VALID_INDEX_TYPES` are defined. This function is used much more frequently by other modules.

Ih that case, I would move *the other way round*, i.e., rename `is_relative_fh_type` to `is_in_valid_relative_index_types` or `is_in_valid_index_types_relative`, and put the three functions together into the `validation` module because they are so very similar.

**Comment by khrapovs** [comment]
> Ih that case, I would move _the other way round_, i.e., rename `is_relative_fh_type` to `is_in_valid_relative_index_types` or `is_in_valid_index_types_relative`, and put the three functions together into the `validation` module because they are so very similar.

Very good. I moved not only `is_in_valid_relative_index_types` and `is_in_valid_absolute_index_types` to `utils/validation/series.py`, but also `RELATIVE_INDEX_TYPES` and `ABSOLUTE_INDEX_TYPES` next to `VALID_INDEX_TYPES` in the same module. I like that!

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great!

I see that, in this newest version, you changed `INDEX_TYPE_LOOKUP` too, but did not address the consequences of that change.

I think we should address those ramifications, before we merge.

There are a number of places in which a check `in INDEX_TYPE_LOOKUP.get(index_type)` occurs, this should be using `is_in_valid_index_types,` no?

**Comment by khrapovs** [comment]
> I see that, in this newest version, you changed `INDEX_TYPE_LOOKUP` too, but did not address the consequences of that change.
> 
> I think we should address those ramifications, before we merge.
> 
> There are a number of places in which a check `in INDEX_TYPE_LOOKUP.get(index_type)` occurs, this should be using `is_in_valid_index_types,` no?

Correct. Thanks for noticing. I have used `is_integer_index` in two places of `test_fh` where `assert isinstance` was testing index type. I have double checked other usages of `INDEX_TYPE_LOOKUP` and do not see anything I should change there.

File: sktime/forecasting/base/tests/test_fh.py
Comment (fkiraly) [inline sktime/forecasting/base/tests/test_fh.py:56]: confused. Instead of this if/else construction, why do you not just write 
`assert is_in_valid_index_types(y.index)`?

**Review by fkiraly — CHANGES_REQUESTED** [review]
See above. Should it not be much shorter than the if/else?
Happy to approve if there is a good reason not to.

File: sktime/forecasting/base/tests/test_fh.py
Comment (khrapovs) [reply sktime/forecasting/base/tests/test_fh.py:56]:
```
assert isinstance(y.index, INDEX_TYPE_LOOKUP.get(index_type))
```
This line checks for a very specific index type, not just the set of valid indices. Since the index that corresponds to `int` key in `INDEX_TYPE_LOOKUP` is very generic `pd.Index` (instead of `pd.Int64Index`) I have to check that it is indeed integer index. Hence, this ugly `if...else`.

Comment (fkiraly) [reply sktime/forecasting/base/tests/test_fh.py:56]: oh, I seel. How about giving `is_in_valid_index_types` an optional second argument, from the lookup dictionary then? Then it checks only for that string.

Comment (khrapovs) [reply sktime/forecasting/base/tests/test_fh.py:56]: Sorry, but at this point I will start to resist. This proposal moves complexity from the test to the source code of the library. We already have a few specific and super simple functions to check for specific index types: `is_integer_index`, `is_in_valid_relative_index_types`, and `is_in_valid_absolute_index_types`. Refactoring those once more to something with optional arguments just to accommodate one single test does not sound very attractive to me.

Comment (fkiraly) [reply sktime/forecasting/base/tests/test_fh.py:56]: > from the test to the source code

Ah, that´s a valid point!
It´s a *test*, not the library code!
Ok, I´m fine with the status quo then.

**Review by fkiraly — APPROVED** [review]
Discussion revealed a good reason to leave as is, which resolves the last change request.

**Pull request merged**
